include/unifex/at_coroutine_exit.hpp

```diff
@@ -0,0 +1,283 @@
+        typename sender_traits<Sender>::template value_types<Variant, Tuple>;
+
+    template <template <typename...> class Variant>
+    using error_types =
+        typename sender_traits<Sender>::template error_types<Variant>;
+
+    static constexpr bool sends_done = false;
+
+    template (typename Receiver)
+      (requires sender_to<Sender, _die_on_done_rec_t<Receiver>>)
+    auto connect(Receiver&& rec) &&
+        noexcept(is_nothrow_connectable_v<Sender, _die_on_done_rec_t<Receiver>>)
+        -> connect_result_t<Sender, _die_on_done_rec_t<Receiver>> {
+      return unifex::connect(
+          (Sender&&) sender_,
+          _die_on_done_rec_t<Receiver>{(Receiver&&) rec});
+    }
+
+    Sender sender_;
+  };
+};
+
+template <typename Sender>
+using _die_on_done_t =
+    typename _die_on_done<remove_cvref_t<Sender>>::type;
+
+struct _die_on_done_fn {
+  template (typename Value)
+    (requires (!detail::_awaitable<Value>) AND sender<Value>)
+  _die_on_done_t<Value> operator()(Value&& value) /*mutable*/
+      noexcept(std::is_nothrow_constructible_v<remove_cvref_t<Value>, Value>) {
+    return _die_on_done_t<Value>{(Value&&) value};
+  }
+
+  template <typename Value>
+  Value&& operator()(Value&& value) const noexcept {
+    return (Value&&) value;
+  }
+};
+
+template <typename... Ts>
+struct _cleanup_task;
+
+template <typename... Ts>
+struct _cleanup_promise : _cleanup_promise_base {
+  template <typename Action>
+  explicit _cleanup_promise(Action&&, Ts&... ts) noexcept
+    : args_(ts...) {}
+
+  _cleanup_task<Ts...> get_return_object() noexcept {
+    return _cleanup_task<Ts...>(
+        coro::coroutine_handle<_cleanup_promise>::from_promise(*this));
+  }
+
+  coro::coroutine_handle<> unhandled_done() noexcept {
+    // Record that we are processing an unhandled done signal. This is checked in
+    // the final_suspend of the cleanup action to know which subsequent continuation
+    // to resume.
+    isUnhandledDone_ = true;
+    // On unhandled_done, run the cleanup action:
+    return coro::coroutine_handle<_cleanup_promise>::from_promise(*this);
+  }
+
+  template <typename Value>
+  decltype(auto) await_transform(Value&& value)
+      noexcept(noexcept(
+          unifex::await_transform(*this, _die_on_done_fn{}((Value&&) value)))) {
+    return unifex::await_transform(*this, _die_on_done_fn{}((Value&&) value));
+  }
+
+  std::tuple<Ts&...> args_;
+};
+
+template <typename... Ts>
+struct [[nodiscard]] _cleanup_task {
+  using promise_type = _cleanup_promise<Ts...>;
+
+  explicit _cleanup_task(coro::coroutine_handle<promise_type> coro) noexcept
+    : continuation_(coro) {}
+
+  _cleanup_task(_cleanup_task&& that) noexcept
+    : continuation_(std::exchange(that.continuation_, {})) {}
+
+  ~_cleanup_task() {
+    UNIFEX_ASSERT(!continuation_);
+  }
+
+  bool await_ready() const noexcept {
+    return false;
+  }
+
+  template <typename Promise>
+  bool await_suspend(coro::coroutine_handle<Promise> parent) noexcept {
```

Comment:
If I understand correctly, each time you co_await a cleanup task it inserts itself in the continuation chain.

The cleanup actions run in *reverse* order, like destructors? I think that is right from the code, and from the behaviour of test_two_cleanup_actions, but the docblock doesn't make that clear.